marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/builtins/steps/mastering/default/merging.sjs

```diff
@@ -40,84 +40,105 @@ function main(content, options) {
+  const thisMatchSummaryURI = content.uri;
   masteringStepLib.checkOptions(null, options, null, requiredOptionProperties);
   const matchSummaryCollection = `datahubMasteringMatchSummary${options.targetEntityType ? `-${options.targetEntityType}`:''}`;
   const collectionQuery = cts.collectionQuery(matchSummaryCollection);
-  const uriRangeQuery = cts.rangeQuery(urisPathReference, '=', uriToTakeActionOn);
-  let relatedMatchSummaries = cts.search(
-    cts.andQuery([
-      uriRangeQuery,
-      collectionQuery
-    ]),
-    ['unfiltered',cts.indexOrder(datahubCreatedOnRef, 'descending')]
-  );
-  let mostRecentMatchSummary = fn.head(relatedMatchSummaries);
-  let matchSummaryJson = mostRecentMatchSummary.toObject();
-  let uriActionDetails = matchSummaryJson.matchSummary.actionDetails[uriToTakeActionOn];
-  // If the action is merge, ensure we create the merge with the most URIs
-  if (uriActionDetails && uriActionDetails.action === 'merge') {
-    const urisQuery = cts.pathRangeQuery('//actionDetails/*/uris', '=', uriActionDetails.uris)
-    const postiveQuery = [
-      cts.jsonPropertyValueQuery('action', 'merge'),
-      urisQuery,
-      collectionQuery
-    ];
-    const otherMergesForURI = cts.search(
-      cts.andNotQuery(
-        cts.andQuery(postiveQuery),
-        uriRangeQuery
-      ),
-      ['unfiltered',cts.indexOrder(datahubCreatedOnRef, 'descending')]
-    ).toArray()
-        // get the actionNode for the URI
-        .map(
-          (result) => result.xpath(`/matchSummary/actionDetails/*[action = 'merge']`).toArray()
-            .filter((actionNode) => cts.contains(actionNode, urisQuery))[0]
-        )
-        // filter out false positives that don't have merge actions for the given URI
-        .filter((actionNode) => !!actionNode)
-        // convert node to JSON object
-        .map((actionNode) => actionNode.toObject());
-    if (otherMergesForURI.some((otherMerge) => otherMerge.uris.length > uriActionDetails.uris.length)) {
-      return Sequence.from([]);
+
+  let returnArray = [];
+
+  let thisMatchSummary = fn.head(cts.search(cts.documentQuery(thisMatchSummaryURI), ['unfiltered']));
+  if (!thisMatchSummary) {
+    return [];
+  }
+  let theseURIsToProcess = thisMatchSummary.xpath('/matchSummary/URIsToProcess').toArray();
+
+  for (uriToProcess of theseURIsToProcess) {
+    // don't process this "URIsToProcess" URI unless we are processing the last matchSummary document
+    // (in URI order) that contains it
+    let lastSummaryWithURI = cts.uris(null, ["descending", "limit=1"],
+      cts.pathRangeQuery("/matchSummary/URIsToProcess", "=" , uriToProcess));
+    if (lastSummaryWithURI != thisMatchSummaryURI) {
+      continue;
     }
-    otherMergesForURI.forEach((actionDetails) => actionDetails.uris.forEach(
-      (uri) => {
-        if (!uriActionDetails.uris.includes(uri)) {
-          uriActionDetails.uris.push(uri);
+
+    const uriRangeQuery = cts.rangeQuery(urisPathReference, '=', uriToProcess);
```

Comment:
My main recommendation so far is to try to break this 70-line stretch of code up into several functions. There are several chunks of processing to work through here - moving each chunk into its own function would allow the function name to document what the chunk is doing, and also what the inputs and outputs are.